martha/martha_v3.js

```diff
@@ -1,36 +1,156 @@
+        port: parsedUrl.port,
+        pathname: `/ga4gh/drs/v1/objects${parsedUrl.pathname}`,
+        search: parsedUrl.search
+    });
 }
 
-function getDataObjectMetadata(dataObjectResolutionUrl, auth, bondProvider) {
-    if (bondProvider === BondProviders.JADE_DATA_REPO) {
-        return apiAdapter.getJsonFrom(dataObjectResolutionUrl, auth);
+/** *************************************************************************************************
+ * Response parsers
+ */
+
+function dosResponseParser (response) {
+    if (response.data_object) {
+        const accessMethods = (response.data_object.urls) ? response.data_object.urls
+            .filter((e) => e.url.startsWith('gs://'))
+            .map((gsUrl) => {
+                return { type: 'gs', access_url: { url: gsUrl.url } };
+            }) : null;
+        return {
+            access_methods: accessMethods,
+            checksums: response.data_object.checksums,
+            created_time: response.data_object.created,
+            mime_type: response.data_object.mimeType,
+            size: response.data_object.size,
+            updated_time: response.data_object.updated,
+        };
     } else {
-        return apiAdapter.getJsonFrom(dataObjectResolutionUrl);
+        throw new FailureResponse(400, `Expected but did not receive properly formatted DOS Response: ${JSON.stringify(response)} `);
     }
 }
 
-function aggregateResponses(responses) {
-    const drsResponse = responses[0];
-    const googleServiceAccount = responses[1];
-    return convertToMarthaV3Response(drsResponse, googleServiceAccount);
+function drsResponseParser (response) {
+    return {
+        access_methods: response.access_methods,
+        checksums: response.checksums,
+        created_time: response.created_time,
+        mime_type: response.mime_type,
+        name: response.name,
+        size: response.size,
+        updated_time: response.updated_time,
+    };
 }
 
-function marthaV3Handler(req, res) {
+/** *************************************************************************************************
+ * Here is where all the logic lives that pairs a particular kind of URI with its
+ * resolving-URL-generating parser, what path to use to make a Bond request for an SA (if any), and
+ * a response parser.
+ */
+
+function determineDrsType (parsedUrl) {
+    if ((parsedUrl.host.toLowerCase() === 'dg.4503') || (parsedUrl.host.toLowerCase() === 'dg.712c')) {
+        return new DrsType(
+            dosCompactUrlGenerator(parsedUrl),
+            false,
+            `${config.bondBaseUrl}/api/link/v1/fence/serviceaccount/key`,
+            dosResponseParser);
+    } else if (parsedUrl.host.toLowerCase().startsWith('dg.')) {
+        return new DrsType(
+            dosCompactUrlGenerator(parsedUrl),
+            false,
+            `${config.bondBaseUrl}/api/link/v1/dcf-fence/serviceaccount/key`,
+            dosResponseParser);
+    } else if (parsedUrl.host.endsWith('.dataguids.org')) {
```

Comment:
I would highly recommend that before the team consider this story complete, that there are mocked end-to-end tests for each branch in this function.  In particular, I do not see any mocked tests for `.dataguids.org` in `martha_v3.test.js` (or elsewhere in the codebase).  I believe we are also missing a test for the `else` branch at the end of this function.  I'm not sure if we think that the current mocked tests that exist in that file cover a significant portion of the control flow that is modified by the behavior that we expect to be different from each server, so I would suggest verifying that too.

Reply:
I was concerned about containing the scope of this PR because there were previous worries about creep.  Therefore, I sought to only retain existing functionality/tests and not add any.  Any additional tests can have their own issue in Jira.